### PR TITLE
✨ amp-story-player: add "autoplay" attribute

### PR DESCRIPTION
**Pull request by proyectoramirez**
This PR introduces 1 attribute to the amp-story-player element: autoplay. Final behaviour is to be defined, but for now, it sets the embed mode of the first story in the player to 3 (PREVIEW), which makes it autoplay.

Interacting with a player on autoplay mode will disable autoplay. For now, changing the embed mode of a laid out story "on the run" is not yet supported, so the player only removes the autoplay attribute.

Partial for #29311

/CC: @newmuis @Enriqe @connielei @bettyashley

### CODE REVIEW

File: src/amp-story-player/amp-story-player-impl.js [hunk 1]
```diff
@@ -905,11 +929,16 @@ export class AmpStoryPlayer {
...
 
   /**
-   * Reacts to touchstart events and caches its coordinates.
+   * Reacts to touchstart events, caches its coordinates,
+   * and ends the autoplay mode if activated.
    * @param {!Event} event
    * @private
    */
   onTouchStart_(event) {
+    if (this.isAutoplaying_) {
```
Comment (Enriqe) [inline src/amp-story-player/amp-story-player-impl.js:938]: Doing this check in `touchStart_` won't work for desktop.

File: src/amp-story-player/amp-story-player-impl.js [hunk 2]
```diff
@@ -352,7 +356,11 @@ export class AmpStoryPlayer {
         story,
         iframe,
         // In case it is the first story, it becomes immediately visibile
-        idx === 0 ? VisibilityState.VISIBLE : VisibilityState.PRERENDER
+        idx === 0 ? VisibilityState.VISIBLE : VisibilityState.PRERENDER,
+        // Autoplaying is applied even if the first story is added dynamically
+        idx === 0 && this.isAutoplaying_
```
Comment (Enriqe) [inline src/amp-story-player/amp-story-player-impl.js:361]: WDYT of making `this.isAutoplaying_ = this.element_.getAttribute('autoplay') && idx === 0` so we don't have to pass an extra argument in all these methods, and just check `this.isAutoplaying_` when it's needed in `this.getEncodedLocation`. 

It's good for code readability to keep the number of arguments in methods low.

File: src/amp-story-player/amp-story-player-impl.js (hunk 1 above)
Comment (proyectoramirez) [reply src/amp-story-player/amp-story-player-impl.js:938]: Where could it go instead?

Comment (proyectoramirez) [reply src/amp-story-player/amp-story-player-impl.js:938]: After looking further at the code, there doesn't seem to be any way to detect any type of user interaction. `onTouchStart_` only works on touchscreen, and `onSelectDocument_` fires any time the story is changed (regardless of whether there has been any interaction)

Maybe I could set up a one-time event listener for click/touch on the element handling this?

File: src/amp-story-player/amp-story-player-impl.js (hunk 2 above)
Comment (proyectoramirez) [reply src/amp-story-player/amp-story-player-impl.js:361]: How about we change `this.isAutoplaying_` from a field to a getter property? I think reading directly the autoplay attribute every time from DOM would require a more robust solution (in the case it's changed programatically), but as it is right now (only making the first story go on preview mode), I think it could work.

I would also suggest `this.getEncodedLocation_` stay as a somewhat pure function, but if that's not the intention, I could put this autoplay embed mode logic there.

File: src/amp-story-player/amp-story-player-impl.js [hunk 3]
```diff
@@ -141,6 +142,9 @@ export class AmpStoryPlayer {
     /** @private {number} */
     this.currentIdx_ = 0;
 
+    /** @private {boolean} */
+    this.isAutoplaying_ = this.element_.getAttribute('autoplay') !== null;
```
Comment (Enriqe) [inline src/amp-story-player/amp-story-player-impl.js:146]:
```suggestion
    this.isAutoplaying_ = this.element_.hasAttribute('autoplay');
```

File: src/amp-story-player/amp-story-player-impl.js (hunk 1 above)
Comment (Enriqe) [reply src/amp-story-player/amp-story-player-impl.js:938]: Yep, I think we need to set up a click listener at the player level, so that we can catch this kind of interactions.

File: src/amp-story-player/amp-story-player-impl.js (hunk 2 above)
Comment (Enriqe) [reply src/amp-story-player/amp-story-player-impl.js:361]: Leaving like it is is fine, we can do a cleanup later if needed.

File: src/amp-story-player/amp-story-player-impl.js (hunk 1 above)
Comment (proyectoramirez) [reply src/amp-story-player/amp-story-player-impl.js]: I tried a new approach on the latest commits. How is it looking?

**Comment by proyectoramirez** [comment]
There is an issue where travis will complain about the player depending on embed-mode.js. Should I replicate that enum?
@Enriqe

**Comment by proyectoramirez** [comment]
Update: I followed @newmuis suggestion and moved `embed-mode.js` out of the `extensions/amp-story/` folder and into the `src/` folder. Is that okay? Or should I move it somewhere more specific? @Enriqe @gmajoulet

File: src/amp-story-player/amp-story-player-impl.js [hunk 4]
```diff
@@ -309,9 +313,25 @@ export class AmpStoryPlayer {
...
+  /**
+   * Sets up a listener disable autoplay mode on
+   * the first user interaction.
+   * @private
+   */
+  setUpAutoplayClickShield_() {
+    this.element_.addEventListener(
+      'click',
+      () => this.toggleAutoplayMode_(false),
+      {once: true}
```
Comment (Enriqe) [inline src/amp-story-player/amp-story-player-impl.js:331]: What would happen in the following case?

player in "preview" -> user clicks -> player exits out of "preview" -> user exits player

i think we would want the player to toggle back to "preview" mode, right? if so, then we would need this listener to persist, and not add/remove on every toggle.

File: src/embed-mode.js [hunk 5]
```diff
@@ -13,8 +13,8 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-import {isEnumValue} from '../../../src/types';
-import {parseQueryString} from '../../../src/url';
+import {isEnumValue} from './types';
+import {parseQueryString} from './url';
 
 /**
  * Embed mode for AMP story.  See ../embed-modes.md for details.
```
Comment (Enriqe) [inline src/embed-mode.js:20]: We'll need to update this path if we do move it.

**Review by Enriqe — APPROVED** [review]
Looks good! 👏

Don't feel strongly about where to move `embed-mode.js`, but if we move it to `src/` we'll probably want to specify it's for `amp-story`, so we would need to change the name to `amp-story-embed-mode.js` or something. I'll defer to @gmajoulet and @newmuis on this one.

File: src/amp-story-player/amp-story-player-impl.js (hunk 4 above)
Comment (proyectoramirez) [reply src/amp-story-player/amp-story-player-impl.js:331]: I see. The click shield already "shows" automatically when the player has the autoplay attribute, so making the listener persist would make it work the same every time.

The only issue is there is no way to re-set the autoplay mode (which I think should be up to the publisher)

File: css/amp-story-player.css [hunk 6]
```diff
@@ -65,3 +65,14 @@ amp-story-player::after {
 .i-amphtml-story-player-loaded::after {
   visibility: hidden;
 }
+
+/** Apply click shield on the player if it is on autoplay mode. */
+.i-amphtml-story-player-loaded[autoplay]::after {
+  all: unset;
```
Comment (Enriqe) [inline css/amp-story-player.css:71]: why is this needed?

**Review by Enriqe — APPROVED** [review]
LGTM! 👏

PTAL @gmajoulet @newmuis

**Review by Enriqe — CHANGES_REQUESTED** [review]
Blocking until we resolve some nuances

File: src/amp-story-player/amp-story-player-impl.js [hunk 7]
```diff
@@ -833,17 +863,23 @@ export class AmpStoryPlayer {
...
+    embedMode = EmbedMode.NOT_EMBEDDED
+  ) {
     const params = dict({
       'amp_js_v': '0.1',
       'visibilityState': visibilityState,
       'origin': this.win_.origin,
       'showStoryUrlInfo': '0',
       'storyPlayer': 'v0',
       'cap': 'swipe',
+      [EmbedModeParam]: embedMode,
```
Comment (Enriqe) [inline src/amp-story-player/amp-story-player-impl.js:882]: I think we need to take into account the case where the publisher already specified an `embedMode`. 

e.g. there are some users of the player already setting the stories in their player like this:
```
<amp-story-player>
  <a href="mystory.com/story.html#embedMode=1"> // We should respect embedMode=1
</amp-story-player>
```

Comment (proyectoramirez) [reply src/amp-story-player/amp-story-player-impl.js:882]: I tried testing this in master (without the changes in this PR) and noticed the current bahavior may be incorrect anyways.

I added the following story: `https://www-usatoday-com.cdn.ampproject.org/c/s/www.usatoday.com/amp-stories/beyonce-rule-the-world/#embedMode=3`

It already has a fragment, as you can see. When parsed with this method, the output is: `https://www-usatoday-com.cdn.ampproject.org/c/s/www.usatoday.com/amp-stories/beyonce-rule-the-world/#embedMode=3?amp_js_v=0.1#visibilityState=prerender&origin=http%3A%2F%2Flocalhost%3A8000&showStoryUrlInfo=0&storyPlayer=v0&cap=swipe`

This has two fragment parameter sections, which is an invalid URL. I am actually not sure how it ends up working, but I think I'm going to end up having to change a significant part of this method.

Comment (proyectoramirez) [reply src/amp-story-player/amp-story-player-impl.js:882]: With further testing, I found that, in that last URL

```
https://www-usatoday-com.cdn.ampproject.org/c/s/www.usatoday.com/amp-stories/beyonce-rule-the-world/#embedMode=3?amp_js_v=0.1#visibilityState=prerender&origin=http%3A%2F%2Flocalhost%3A8000&showStoryUrlInfo=0&storyPlayer=v0&cap=swipe
```

The hash part is considered to be:

```
#embedMode=3?amp_js_v=0.1#visibilityState=prerender&origin=http%3A%2F%2Flocalhost%3A8000&showStoryUrlInfo=0&storyPlayer=v0&cap=swipe
```

With the `parseQueryString` function from `src/url`, this is parsed as:

```
{
  cap: "swipe"
  embedMode: "3?amp_js_v=0.1#visibilityState=prerender"
  origin: "http://localhost:8000"
  showStoryUrlInfo: "0"
  storyPlayer: "v0"
}
```

Notice how the `amp_js_v` query and the `visibilityState` fragment are all grouped under the `embedMode` parameter. This still results in the appropiate embedMode being applied, as the `parseEmbedMode` function [here](https://github.com/ampproject/amphtml/blob/2fb6f8e0d7ebb0a0944faf91c1a8f306db8d9146/extensions/amp-story/1.0/embed-mode.js#L91) uses `parseInt`, which parses integers until non-integers are found, still giving 3 as a result, appropiately.

I have also found that some stories do actually break with an incorrect URL (missing `amp_js_v` query), like [this one](https://www-washingtonpost-com.cdn.ampproject.org/v/s/www.washingtonpost.com/graphics/2019/lifestyle/travel/amp-stories/a-locals-guide-to-what-to-eat-and-do-in-miami/#embedMode=3?amp_js_v=0.1#visibilityState=prerender&origin=http%3A%2F%2Flocalhost%3A8000&showStoryUrlInfo=0&storyPlayer=v0&cap=swipe).

I'm fixing this in the next commits.

Comment (proyectoramirez) [reply src/amp-story-player/amp-story-player-impl.js]: After I uploaded the "fixes", I noticed a test for this started failing. Turns out it was checking against the following URL:
```
https://www-washingtonpost-com.cdn.ampproject.org/v/s/www.washingtonpost.com/graphics/2019/lifestyle/travel/amp-stories/a-locals-guide-to-what-to-eat-and-do-in-new-york-city/?testParam=true#myhash=hashValue&amp_js_v=0.1#visibilityState=visible&origin=http%3A%2F%2Flocalhost%3A9876&showStoryUrlInfo=0&storyPlayer=v0&cap=swipe&embedMode=0
```
This is also incorrect (has 2 hashes), so I changed it.

File: src/amp-story-player/amp-story-player-impl.js [hunk 8]
```diff
@@ -833,29 +856,48 @@ export class AmpStoryPlayer {
...
+      ...originalFragments
+    } = parseQueryString(originalFragmentString);
+
+    const newFragments = {
       'visibilityState': visibilityState,
       'origin': this.win_.origin,
       'showStoryUrlInfo': '0',
       'storyPlayer': 'v0',
       'cap': 'swipe',
+      [EmbedModeParam]:
```
Comment (Enriqe) [inline src/amp-story-player/amp-story-player-impl.js:883]: Can we just not add the `embedMode` parameter if not provided?

File: src/amp-story-player/amp-story-player-impl.js [hunk 9]
```diff
@@ -309,9 +315,21 @@ export class AmpStoryPlayer {
     this.initializeIframes_();
     this.initializeButton_();
     this.signalReady_();
+    this.setUpAutoplayClickShield_();
```
Comment (Enriqe) [inline src/amp-story-player/amp-story-player-impl.js:318]: Can we set this up only when the `preview` attribute is present?

File: src/amp-story-player/amp-story-player-impl.js [hunk 10]
```diff
@@ -833,29 +856,48 @@ export class AmpStoryPlayer {
    * Gets encoded url for player usage.
    * @param {string} href
    * @param {VisibilityState=} visibilityState
+   * @param {?EmbedMode=} embedMode If not provided, it will keep
```
Comment (Enriqe) [inline src/amp-story-player/amp-story-player-impl.js:859]:
```suggestion
   * @param {?EmbedMode=} embedMode If provided, it will keep
```

File: src/amp-story-player/amp-story-player-impl.js [hunk 11]
```diff
@@ -833,29 +856,48 @@ export class AmpStoryPlayer {
...
+    } = parseQueryString(originalFragmentString);
+
+    const newFragments = {
       'visibilityState': visibilityState,
       'origin': this.win_.origin,
       'showStoryUrlInfo': '0',
       'storyPlayer': 'v0',
       'cap': 'swipe',
+      [EmbedModeParam]:
+        embedMode ?? originalEmbedMode ?? EmbedMode.NOT_EMBEDDED,
```
Comment (Enriqe) [inline src/amp-story-player/amp-story-player-impl.js:884]: Referencing my comment above, this could be something like:

```
const newFragments = { ... }
const currentEmbedMode = embedModeOverride || originalEmbedMode;
if (currentEmbedMode) { newFragments[embedMode] = currentEmbedMode }

File: src/amp-story-player/amp-story-player-impl.js (hunk 8 above)
Comment (proyectoramirez) [reply src/amp-story-player/amp-story-player-impl.js]: Done in latest commit. No default is provided if this ends up being `undefined`, which is later skipped in `serializeQueryString`.

File: src/amp-story-player/amp-story-player-impl.js (hunk 9 above)
Comment (proyectoramirez) [reply src/amp-story-player/amp-story-player-impl.js]: Done!

File: src/amp-story-player/amp-story-player-impl.js (hunk 11 above)
Comment (proyectoramirez) [reply src/amp-story-player/amp-story-player-impl.js]: I originally thought about this, but since `undefined` parameters are skipped in `serializeQueryString`, we could also only remove the last `?? EmbedMode.NOT_EMBEDDED`. Would it be better to be more explicit?

File: src/amp-story-player/amp-story-player-impl.js (hunk 7 above)
Comment (Enriqe) [reply src/amp-story-player/amp-story-player-impl.js]: This is great! Thanks for catching and investigating this so thoroughly. 👏👏👏

**Review by Enriqe — APPROVED** [review]
Great work! 👏👏👏

File: css/amp-story-player.css [hunk 12]
```diff
@@ -65,3 +65,14 @@ amp-story-player::after {
 .i-amphtml-story-player-loaded::after {
   visibility: hidden;
 }
+
+/** Apply click shield on the player if it is on autoplay mode. */
+.i-amphtml-story-player-loaded[autoplay]::after {
+  all: initial;
```
Comment (gmajoulet) [inline css/amp-story-player.css:71]: Not needed?

File: extensions/amp-story/1.0/test/test-amp-story-store-service.js [hunk 13]
```diff
@@ -20,7 +20,7 @@ import {
   StateProperty,
   UIType,
 } from '../amp-story-store-service';
-import {EmbedMode, EmbedModeParam} from '../embed-mode';
+import {EmbedMode, EmbedModeParam} from '../../../../src/amp-story-embed-mode';
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/test/test-amp-story-store-service.js:23]: Why are we moving this file? It clearly belongs in the `/amp-story` folder. We don't even have OWNERS rights on the `/src` folder.

The Player is one client for AMP Story and should do like other clients that have to duplicate the EmbedMode enum. I'm fine with either duplication or importing the enum from `/amp-story` (and whitelisting the import so the compiler doesn't complain), but moving the file is confusing.

File: src/amp-story-player/amp-story-player-impl.js [hunk 14]
```diff
@@ -309,9 +315,25 @@ export class AmpStoryPlayer {
...
   }
 
+  /**
+   * Sets up a listener disable autoplay mode on
+   * the first user interaction.
+   * @private
+   */
+  setUpAutoplayClickShield_() {
+    this.element_.addEventListener('click', () =>
+      this.toggleAutoplayMode_(false)
```
Comment (gmajoulet) [inline src/amp-story-player/amp-story-player-impl.js:333]: This code will run on every click within the player. Should we unlisten when autoplay is off, and listen again when autoplay is on?
There would be no way to set it back on otherwise, as this would set it back to off.

File: src/amp-story-player/amp-story-player-impl.js [hunk 15]
```diff
@@ -793,11 +818,13 @@ export class AmpStoryPlayer {
    * @param {!Element} story
    * @param {!Element} iframe
    * @param {!VisibilityState} visibilityState
+   * @param {EmbedMode=} embedMode
```
Comment (gmajoulet) [inline src/amp-story-player/amp-story-player-impl.js:821]: Nit: can't be null

```suggestion
   * @param {!EmbedMode=} embedMode
```

File: src/amp-story-player/amp-story-player-impl.js [hunk 16]
```diff
@@ -793,11 +818,13 @@ export class AmpStoryPlayer {
...
    * @private
    */
-  layoutIframe_(story, iframe, visibilityState) {
+  layoutIframe_(story, iframe, visibilityState, embedMode) {
```
Comment (gmajoulet) [inline src/amp-story-player/amp-story-player-impl.js:824]: If the param is optional you should specify a default value

```suggestion
  layoutIframe_(story, iframe, visibilityState, embedMode = undefined) {
```

File: src/amp-story-player/amp-story-player-impl.js [hunk 17]
```diff
@@ -833,29 +860,47 @@ export class AmpStoryPlayer {
    * Gets encoded url for player usage.
    * @param {string} href
    * @param {VisibilityState=} visibilityState
+   * @param {?EmbedMode=} embedMode If provided, it will override
```
Comment (gmajoulet) [inline src/amp-story-player/amp-story-player-impl.js:863]: Re: !EmbedMode= and default value

File: src/amp-story-player/amp-story-player-impl.js (hunk 14 above)
Comment (Enriqe) [reply src/amp-story-player/amp-story-player-impl.js:333]: Given that the iframe will consume the click when the "click shield" is not applied, I don't think it's necessary to detach/attach this listener on every toggle, which is why I recommended doing it this way. But unlistening/listening on toggle sgtm!

File: css/amp-story-player.css (hunk 12 above)
Comment (proyectoramirez) [reply css/amp-story-player.css:71]: It is, since there are styles applied to that `::after` element previously, including an animation that makes it spin 😵

**Review by Enriqe — APPROVED** [review]

File: test/unit/test-amp-story-player.js [hunk 18]
```diff
@@ -745,5 +748,35 @@ describes.realWin('AmpStoryPlayer', {amp: false}, (env) => {
...
+    it('sets the Embed Mode of only the first story to PREVIEW when in autoplay mode', async () => {
+      const playerEl = win.document.createElement('amp-story-player');
+      playerEl.setAttribute('autoplay', '');
+      appendStoriesToPlayer(playerEl, 3);
+
+      const player = new AmpStoryPlayer(win, playerEl);
+      await player.load();
+      await nextTick();
+
+      const storyIframes = playerEl.shadowRoot.querySelectorAll('iframe');
```
Comment (Enriqe) [inline test/unit/test-amp-story-player.js:761]:
```suggestion
      const storyIframes = playerEl.querySelectorAll('iframe');
```
